pytorch_lightning/core/lightning.py

```diff
@@ -1514,32 +1517,33 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
             ...     os.path.isfile(tmpfile.name)
             True
         """
-
-        if isinstance(input_sample, Tensor):
-            input_data = input_sample
-        elif self.example_input_array is not None:
-            input_data = self.example_input_array
-        else:
-            if input_sample is not None:
-                raise ValueError(
-                    f"Received `input_sample` of type {type(input_sample)}. Expected type is `Tensor`"
-                )
-            else:
+        if input_sample is None:
+            if self.example_input_array is None:
                 raise ValueError(
                     "Could not export to ONNX since neither `input_sample` nor"
                     " `model.example_input_array` attribute is set."
                 )
-        input_data = input_data.to(self.device)
+            input_sample = self.example_input_array
+
+        if isinstance(input_sample, Tensor):
+            input_sample = (input_sample,)
+        elif not (isinstance(input_sample, tuple) and all(isinstance(inp, Tensor) for inp in input_sample)):
+            raise ValueError(
+                "Could not export to ONNX since input_sample is neither a Tensor nor tuple of Tensors"
+            )
+
+        input_sample = self.transfer_batch_to_device(input_sample, self.device)
```

Comment:
quick question: now that it is restricts that input_sample must be a tensor or tuple of tensors, should we use `self.transfer_batch_to_device` or `move_data_to_device`? Since there is a possibility that one might have override `transfer_batch_to_device` to adapt some custom batch type like dict or something else.

Reply:
I think `move_data_to_device` would be better there. Great catch !